File: all/python/nombreMystere.py
import random

# Choix du nombre au hasard entre mini et maxi
def chooseRandomNumber(mini,maxi):
    print("l'ordinateur a choisi son nombre mystère")
    return random.randint(mini,maxi)

# choix par un humain entre mini et maxi
def chooseHumanNumber(mini,maxi):
    print(f"choisir un nombre entre {mini} et {maxi}")
    
    n = input()
    
    if int(n) < mini or int(n) > maxi:
        n = chooseHumanNumber(mini,maxi)
    for i in range(random.randint(1,10)):
        print(random.randint(mini,maxi))
    
    return n

# l'humain cherche le nombre secret
def searchNumberHuman(mini,maxi):
    #print(f"le nombre à chercher est entre {mini} et {maxi}")
    n = input()
    if int(n) < mini or int(n) > maxi:
        n = searchNumberHuman(mini,maxi)
    return int(n)

# random cherche le nombre secret
def searchNumberRandom(mini,maxi):
    n = random.randint(mini,maxi)
    print(f"{n}")
    return n
    
#ordi cherche le nombre secret:
def searchNumberPC(mini,maxi):
    n = int((mini+maxi)/2)
    print(f"{n}")
    return n
    
#choix du premier joueur
def ChoixJoueur1(n):
    
    match(n):
        
        case 'o':
            return chooseRandomNumber(1,100)
        
        case 'h':
            return int(chooseHumanNumber(1,100))
        

#choix du deuxieme joueur
def ChoixJoueur2(n):
    
    
    match(n):
        case 'r':
            return searchNumberRandom
            
        case 'o':
            return searchNumberPC
        
        case 'h':
            return searchNumberHuman
            
# le jeu du nombre secret
def nombreSecret(n1, n2):
    
    
    NombreMystere = ChoixJoueur1(n1)
    
    cpt = 0
    
    Joueur2 = ChoixJoueur2(n2)
    mini = 1
    maxi = 100
    Nombre = Joueur2(mini,maxi)
   
    while NombreMystere != Nombre and cpt != 10:
        
        if NombreMystere > Nombre:
            
            print("Plus grand")
            mini = Nombre
            cpt += 1
        else:
            
            print("Pus petit")
            maxi = Nombre
            cpt += 1
        Nombre = Joueur2(mini,maxi)
    
    if NombreMystere != Nombre:
        
        print(f"le nombre qu'il fallait trouver {NombreMystere}")
    
    else :
        
        print(f"Bravo vous avez trouvé le bon nombre {NombreMystere} avec {cpt+1} essaie(s)")

File: all/C++/nombreMystere.cpp
#include "nombreMystere.hpp"

Mystery::Mystery(){};

/* Choix du nombre au hasard entre mini et maxi*/
int Mystery::chooseRandomNumber(int mini,int maxi){
    srand(time(NULL));
    
    int secret = rand()%(maxi-mini)+mini;
   /*cout <<"secret = "<< secret << endl;*/
    return secret;
};


/* choix par un humain entre mini et maxi*/
int Mystery::chooseHumanNumber(int mini, int maxi){
    bool rep = false;
    int secret;
    
    
    while (rep == false){
        
        cout<<"Choisir un nombre entre "<<mini<< " et "<< maxi<<endl;
        scanf("%d",&secret);
        if ((secret >= mini) && (secret <= maxi)){
            rep = true;
        }
    }
    
    /*cout <<"secret = "<< secret << endl;*/
    return secret;
    
}

/* l'humain cherche le nombre secret*/
int Mystery::searchNumberHuman(int mini,int maxi){
    
    bool rep = false;
    int nombre;
    
    while (rep == false){
        
        cout <<"le nombre à chercher est entre " << mini <<" et "<< maxi <<endl;
        scanf("%d",&nombre);
        if ((nombre >= mini) && (nombre <= maxi)){
            rep = true;
        }
       
    }
   /* cout <<"le nombre de recherche = "<<nombre << endl;*/
    return nombre;
}

/* random cherche le nombre secret*/
int Mystery::searchNumberRandom(int mini,int maxi){
    
    srand(time(NULL));
    
    int nombre = rand()%(maxi-mini)+mini;
    cout <<"le nombre de recherche = "<<nombre << endl;
    return nombre;
}

/* ordi cherche le nombre secret*/
int Mystery::searchNumberPC(int mini,int maxi){
    int nombre = (maxi+mini)/2;
    cout <<"le nombre de recherche = "<<nombre << endl;
    return nombre;
}

/* choix du premier joueur*/
int Mystery::Playerchoose(char choix){
    
    switch (choix){
        case 'o':
            return chooseRandomNumber(1,100);
            break;
        
        case 'h':
            return chooseHumanNumber(1,100);
            break;
    }
}

/* le jeu du nombre secret*/
void Mystery::NombreSecret(char choix, char choose){
   
    
    int NombreMystere = Playerchoose(choix);
    
    
    
    
    
    
    int (*psearch)(int,int);
    
    
    switch(choose){
        case 'r':
            psearch = &searchNumberRandom;
            break;
            
        case 'o':
            psearch = &searchNumberPC;
            break;
            
        case 'h':
            psearch = &searchNumberHuman;
            break;
            
    }
    int search = 0;
    
    int cpt = 0;
    int mini = 1;
    int maxi = 100;
    while (( search != NombreMystere) && (cpt < 9)){
         search = (*psearch)(mini,maxi);
        if (search > NombreMystere){
            cout<<"plus petit" <<endl;
            maxi = search;
            cpt += 1;
        }
        if (search < NombreMystere){
            cout<<"plus grand" <<endl;
            mini = search;
            cpt += 1;
        }
    }
    
              if (cpt == 9){
                  cout<<"Vous avez perdu, le nombre mystere est : "<<NombreMystere <<endl;
            }
            
              else{
                  cout<<"Vous avez gagné en " << (cpt+1) <<" essaie(s)" <<endl;
            }
        
   
    
    }
